src/main/sensors/compass.c

```diff
@@ -388,31 +398,233 @@ uint32_t compassUpdate(timeUs_t currentTimeUs)
+ *  zn(1) = 1 / ( lambda + gyro(i,3)*dP(2) - gyro(i,2)*dP(3) );
+ *  g(1) = zn(1) * dP(1);
+ *  g(2) = zn(1) * dP(2);
+ *  g(3) = zn(1) * dP(3);
+ *  b(1) = b(1) - e(1) * g(1);
+ *  b(2) = b(2) - e(1) * g(2);
+ *  b(3) = b(3) - e(1) * g(3);
+ *  P(1,1) = P(1,1) - g(1) * dP(1);
+ *  P(2,1) = P(2,1) - g(2) * dP(1);
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+ *
+ *  // iteration 2
+ *  dP = [P(3,1)*gyro(i,1) - P(1,1)*gyro(i,3); ...
+ *        P(3,2)*gyro(i,1) - P(2,1)*gyro(i,3); ...
+ *        P(3,3)*gyro(i,1) - P(3,1)*gyro(i,3)];
+ *  zn(2) = 1/(lambda - gyro(i,3)*dP(1) + gyro(i,1)*dP(3));
+ *  g(1) = zn(2) * dP(1);
+ *  g(2) = zn(2) * dP(2);
+ *  g(3) = zn(2) * dP(3);
+ *  b(1) = b(1) - e(2) * g(1);
+ *  b(2) = b(2) - e(2) * g(2);
+ *  b(3) = b(3) - e(2) * g(3);
+ *  P(1,1) = P(1,1) - g(1) * dP(1);
+ *  P(2,1) = P(2,1) - g(2) * dP(1);
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+ *
+ *  // iteration 3
+ *  dP = [P(1,1)*gyro(i,2) - P(2,1)*gyro(i,1); ...
+ *        P(2,1)*gyro(i,2) - P(2,2)*gyro(i,1); ...
+ *        P(3,1)*gyro(i,2) - P(3,2)*gyro(i,1)];
+ *  zn(3) = 1/(lambda + gyro(i,2)*dP(1) - gyro(i,1)*dP(2));
+ *  g(1) = zn(3) * dP(1);
+ *  g(2) = zn(3) * dP(2);
+ *  g(3) = zn(3) * dP(3);
+ *  b(1) = b(1) - e(3) * g(1);
+ *  b(2) = b(2) - e(3) * g(2);
+ *  b(3) = b(3) - e(3) * g(3);
+ *  P(1,1) = P(1,1) - g(1) * dP(1);
+ *  P(2,1) = P(2,1) - g(2) * dP(1);
+ *  P(2,2) = P(2,2) - g(2) * dP(2);
+ *  P(3,1) = P(3,1) - g(3) * dP(1);
+ *  P(3,2) = P(3,2) - g(3) * dP(2);
+ *  P(3,3) = P(3,3) - g(3) * dP(3);
+
+ *  zn = zn * lambda;
+ *  P = P / lambda;
+ *  lambda = lambda_min + (1 - lambda_min) * ( zn.' * zn ) / 3.0;
+ */
+
+void compassBiasEstimatorInit(compassBiasEstimator_t *cBE, const float lambda_min, const float p0)
+{
+    compassBiasEstimatorUpdate(cBE, lambda_min, p0);
+    
+    cBE->lambda = lambda_min;
+
+    // don't know if it is necessary to initialize those as 0
+    for (uint8_t i = 0; i < 3; i++) {
+        cBE->b[i] = 0.0f;
+        for (uint8_t j = 0; j < 3; j++) {
+            if (i != j) {
+                cBE->P[i][j] = 0.0f;
+            }
+        }
+    }  
+}
+
+void compassBiasEstimatorUpdate(compassBiasEstimator_t *cBE, const float lambda_min, const float p0)
+{
+    cBE->lambda_min = lambda_min;
+    cBE->p0 = p0;
+
+    // update diagonal entries for faster convergence
+    for (uint8_t i = 0; i < 3; i++) {
+        for (uint8_t j = 0; j < 3; j++) {
+            if (i == j) {
+                cBE->P[i][j] = cBE->p0;
+            }
+        }
+    } 
+}
+
+void compassBiasEstimatorApply(compassBiasEstimator_t *cBE, float *mag, const float *dmag, const float *gyro)
+{
+    const float e[3] = { dmag[0] + gyro[2] * ( cBE->b[1] - mag[1] ) - gyro[1] * ( cBE->b[2] - mag[2] ),
```

Comment:
(vectorCrossProduct is available. But `vector.h` interface may be unstable now, so minor refactoring + rebase may be necessary)

Maybe  add comment about intended operation here